### PR TITLE
Create and use a custom distroless-base as base for distroless variants

### PR DESCRIPTION
**Pull request by davidjumani**
# Description

Create a custom `distroless-base` and `distroless-base-with-binaries` image to be used as the base for all distroless image variants. This simplifies the distroless dockerfiles

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/distroless/Dockerfile.binaries [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# The base image should be $(GLOO_DISTROLESS_BASE_IMAGE)
+ARG BASE_IMAGE
+
+FROM busybox:1.35.0-uclibc as busybox
```
Comment (ben-taussig-solo) [inline projects/distroless/Dockerfile.binaries:4]: I think for consistency we should define this image in the Makefile. I'm a little unsure of the naming, but I think it probably still makes sense to call it a donor image

File: Makefile [hunk 2]
```diff
@@ -373,6 +376,36 @@ ifeq ($(USE_SILENCE_REDIRECTS), true)
...
+		--build-arg DONOR_IMAGE=$(DONOR_IMAGE) \
+		--build-arg BASE_IMAGE=$(DISTROLESS_BASE_IMAGE) \
+		--build-arg GOARCH=$(GOARCH) \
+		-t $(GLOO_DISTROLESS_BASE_IMAGE) $(QUAY_EXPIRATION_LABEL) $(STDERR_SILENCE_REDIRECT)
+
+$(DISTROLESS_OUTPUT_DIR)/Dockerfile.binaries: $(DISTROLESS_DIR)/Dockerfile.binaries
+	mkdir -p $(DISTROLESS_OUTPUT_DIR)
+	cp $< $@
+
+.PHONY: distroless-with-binaries-docker
```
Comment (ben-taussig-solo) [inline Makefile:403]: Nit: how do you feel about `distroless-with-utils-docker`? maybe just `distroless-utils-docker`? I think `distroless-with-binaries` is slightly unclear about which binaries are included and for what purpose

File: Makefile [hunk 3]
```diff
@@ -78,7 +78,11 @@ TEST_ASSET_DIR := $(ROOTDIR)/_test
 
 # Use a distroless debian variant that is in sync with the ubuntu version used for envoy
 # https://github.com/solo-io/envoy-gloo-ee/blob/main/ci/Dockerfile#L7 - check /etc/debian_version in the ubuntu version used
-DISTROLESS_BASE_IMAGE ?= gcr.io/distroless/base-debian11:latest
+PACKAGE_DONOR_IMAGE ?= debian:11
+BINARY_DONOR_IMAGE ?= busybox:1.35.0-uclibc
+DISTROLESS_BASE_IMAGE ?= gcr.io/distroless/base-debian11:latest-amd64
+GLOO_DISTROLESS_BASE_IMAGE ?= $(IMAGE_REGISTRY)/distroless-base:$(VERSION)
+GLOO_DISTROLESS_BASE_WITH_BINARIES_IMAGE ?= $(IMAGE_REGISTRY)/distroless-base-with-binaries:$(VERSION)
```
Comment (ben-taussig-solo) [inline Makefile:85]: Nit: I think we could benefit from adding comments to clarify the purpose of these images
```suggestion
# used to install ca-certificates in GLOO_DISTROLESS_BASE_IAMGE
PACKAGE_DONOR_IMAGE ?= debian:11
# harvested for utility binaries (sh, wget, sleep, nc, echo, ls, cat, vi)
# in GLOO_DISTROLESS_BASE_WITH_BINARIES_IMAGE
BINARY_DONOR_IMAGE ?= busybox:1.35.0-uclibc
# genuine base image for GLOO_DISTROLESS_BASE_IMAGE and
# GLOO_DISTROLESS_BASE_WITH_BINARIES_IMAGE
DISTROLESS_BASE_IMAGE ?= gcr.io/distroless/base-debian11:latest-amd64
# DISTROLESS_BASE_IMAGE, with ca-certificates installed
GLOO_DISTROLESS_BASE_IMAGE ?= $(IMAGE_REGISTRY)/distroless-base:$(VERSION)
# DISTROLESS_BASE_IMAGE, with utility binaries (sh, wget, sleep, nc, echo, ls, cat, vi)
GLOO_DISTROLESS_BASE_WITH_BINARIES_IMAGE ?= $(IMAGE_REGISTRY)/distroless-base-with-binaries:$(VERSION)
```

File: projects/distroless/Dockerfile.binaries (hunk 1 above)
Comment (ben-taussig-solo) [reply projects/distroless/Dockerfile.binaries]: change at https://github.com/solo-io/gloo/pull/9286/commits/0026a157031e148e3954a16b198969e5bd0377dc looks good! resolving this comment

File: changelog/v1.17.0-beta16/gloo-distroless-base.yaml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+changelog:
+- type: NON_USER_FACING
+  description: Create custom distroless images that contain ca-certificates and the binaries required for the probes / lifecycle hooks. This image can be used as the base for all distroles image variants.
```
Comment (bewebi) [inline changelog/v1.17.0-beta16/gloo-distroless-base.yaml:3]:
```suggestion
  description: Create custom distroless images that contain ca-certificates and the binaries required for the probes / lifecycle hooks. This image can be used as the base for all distroless image variants.
```

File: projects/distroless/Dockerfile.binaries [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+FROM $BASE_IMAGE
+
+# Copy over the required binaries for probes and lifecycle hooks
+COPY --from=donor /bin/sh /bin/sh
+COPY --from=donor /bin/wget /bin/wget
+COPY --from=donor /bin/sleep /bin/sleep
+COPY --from=donor /bin/nc /bin/nc
+COPY --from=donor /bin/echo /bin/echo
+# Copy over these binaries for basic debugging. Can be removed if needed
```
Comment (bewebi) [inline projects/distroless/Dockerfile.binaries:15]: Removed in what sense? Do we mean that a future commit might remove these or that they can be commented out for one-off builds?
If the former, would it make sense to allow this to be toggled based on an argument, or have separate debug and regular Dockerfiles?

**Review by bewebi — COMMENTED** [review]
Looks good
One typo, one question/suggestion

File: Makefile (hunk 3 above)
Comment (davidjumani) [reply Makefile]: Good idea! Updated

File: projects/distroless/Dockerfile.binaries (hunk 5 above)
Comment (davidjumani) [reply projects/distroless/Dockerfile.binaries]: Updated the comment. They are non-essential and can be removed later on. I don't believe we need a toggle as I plan to remove them after a couple of releases

File: Makefile (hunk 2 above)
Comment (davidjumani) [reply Makefile]: Updated

**Review by bewebi — APPROVED** [review]

File: Makefile (hunk 2 above)
Comment (ben-taussig-solo) [reply Makefile]: Looks good, thanks!

**Comment by davidjumani** [comment]
kick bulldozer

**Pull request merged**